File: C++/Day05.cpp
#ifdef DAY05

#include "AH.h"

namespace Day05
{

	typedef std::pair<uint64_t, uint64_t> Range;

	std::vector<uint64_t> parseSeeds(std::string s)
	{
		std::vector<uint64_t> ints;

		const auto ss = s.substr(7);
		const auto sis = AH::Split(ss, ' ');

		for (auto si : sis) {
			ints.push_back(AH::stoui64(si));
		}

		return ints;
	}

	std::list<Range> parseSeedsToRanges(std::string s)
	{
		std::list<Range> ranges;

		const auto ss = s.substr(7);
		const auto sis = AH::Split(ss, ' ');

		for (size_t i = 0; i < sis.size(); i += 2) {
			ranges.emplace_back(AH::stoui64(sis[i]), AH::stoui64(sis[i+1]));
		}

		return ranges;
	}

	struct mapping
	{
		uint64_t r;
		uint64_t s;
		uint64_t t;

		mapping(std::string & s);
	};

	mapping::mapping(std::string & st)
	{
		const auto sis = AH::Split(st, ' ');
		r = AH::stoui64(sis[0]);
		s = AH::stoui64(sis[1]);
		t = AH::stoui64(sis[2]);
	}

	std::vector<mapping> parseMapping(const std::string s)
	{
		std::vector<mapping> ms;
		const int hdr = s.find(":");
		const std::string parsed = s.substr(hdr+1);
		const auto ps = AH::Split(parsed, '|');

		for (auto p : ps) {
			if (p.size() == 0) continue;
			mapping m(p);
			ms.push_back(m);
		}

		return ms;
	}

	uint64_t applyMappingsToInt(const uint64_t n, const std::vector<mapping> ms)
	{
		for (auto m :  ms) {
			if ((n - m.s) < m.t)
				return m.r + (n - m.s);
		}
		return n;
	}

	void applyMappingToRanges(std::list<Range> & pre, const std::vector<mapping> ms)
	{
		std::list<Range> post; // done, we pass these onto the next mapping group
		std::list<Range> next; // can still be affected by a mapping in this group
		for (auto m : ms) {
			while (pre.size() > 0) {
				auto r = pre.front();
				pre.pop_front();
				// if there is no overlap - this range remains unchanged
				// so stick it unchanged into next since another interval could affect it
				if ( (r.first >= (m.s + m.t)) || (m.s >= (r.first + r.second)) ) {
					next.push_back(r);
				} else {
					// the interval of overlap
					uint64_t overlap_lo = std::max(r.first, m.s);
					uint64_t overlap_hi = std::min(r.first + r.second, m.s + m.t);

					// there is an untouched interval at the start...
					if (overlap_lo > r.first) {
						// ...so stick that on the end of the list to process
						next.emplace_back(r.first, overlap_lo - r.first);
					}

					// the region of overlap is reallocated, and not touched again in this group
					post.emplace_back(m.r + overlap_lo - m.s, overlap_hi - overlap_lo);

					// there is an untouched interval at the end...
					if (overlap_hi < r.first + r.second) {
						// ... so stick that on the end of the list to process
						next.emplace_back(overlap_hi, r.first + r.second - overlap_hi);
					}
				}
			}
			// these are the intervals that could still be affected by mappings in this group
			pre = next;
			next.clear();
		}

		pre.insert(pre.end(), post.begin(), post.end());

		return;
	}

	int Run(const std::string& filename)
	{
		const auto inputLines = AH::ReadTextFile(filename);
		const auto ss = AH::ParseLineGroups(inputLines, '|');

		std::vector<std::vector<mapping>> mss;
		for (size_t i = 1; i < ss.size(); ++i) {
			mss.push_back(parseMapping(ss[i]));
		}

		const auto seeds = parseSeeds(ss[0]);

		uint64_t part1 = std::numeric_limits<uint64_t>::max();
		for (auto seed : seeds) {
			for (auto ms : mss) {
				seed = applyMappingsToInt(seed, ms);
			}
			part1 = std::min(part1, seed);
		}

		auto seed_ranges = parseSeedsToRanges(ss[0]);
		for (auto ms : mss) {
			applyMappingToRanges(seed_ranges, ms);
		}

		uint64_t part2 = std::numeric_limits<uint64_t>::max();
		for (auto r : seed_ranges) {
			part2 = std::min(r.first, part2);
		}

		AH::PrintSoln(5, part1, part2);

		return 0;
	}

}

#endif

File: Go/Day05.go
package main

import AH "./adventhelper"

import (
	"math"
	"strconv"
	"strings"
)

type Mapping struct {
	r, s, t int
}

type Range struct {
	s, w int
}

func parseSeeds(s string) []int {
	is := []int{}
	ss := strings.Split(s[7:], " ")

	for _, s := range ss {
		i, _ := strconv.Atoi(s)
		is = append(is, i)
	}

	return is
}

func parseSeedsToRanges(s string) []Range {
	rs := []Range{}

	ss := strings.Split(s[7:], " ")

	for i := 0; i < len(ss); i += 2 {
		start, _ := strconv.Atoi(ss[i])
		width, _ := strconv.Atoi(ss[i+1])
		rs = append(rs, Range{start, width})
	}

	return rs
}

func parseMapping(s string) []Mapping {
	ms := []Mapping{}

	ss := strings.Split(s, "|")
	for _, s := range ss {
		if AH.ContainsChar(s, ':') {
			continue
		}
		p := strings.Split(s, " ")
		r, _ := strconv.Atoi(p[0])
		s, _ := strconv.Atoi(p[1])
		t, _ := strconv.Atoi(p[2])
		ms = append(ms, Mapping{r,s,t})
	}

	return ms
}

func applyMappingToRanges(rs []Range, ms []Mapping) []Range {
	post := []Range{}
	next := []Range{}

	for _, m := range ms {
		for ; 0 < len(rs); {
			r := rs[0]
			rs = rs[1:]
			if (r.s >= (m.s + m.t)) || (m.s >= (r.s + r.w)) {
				next = append(next, r)
			} else {
				overlap_s := AH.Max(r.s, m.s)
				overlap_e := AH.Min(r.s + r.w, m.s + m.t)

				if (overlap_s > r.s) {
					next = append(next, Range{r.s, overlap_s - r.s})
				}

				post = append(post, Range{overlap_s + m.r - m.s, overlap_e - overlap_s})

				if (overlap_e < r.s + r.w) {
					next = append(next, Range{overlap_e, r.s + r.w - overlap_e})
				}
			}
		}
		rs = next
		next = []Range{}
	}

	rs = append(rs, post...)

	return rs
}

func applyMappingToInt(n int, ms []Mapping) int {
	for _, m := range ms {
		if (n >= m.s) && ((n - m.s) < m.t) {
			return m.r + (n - m.s)
		}
	}
	return n
}

func main() {
	inputLines, _ := AH.ReadStrFile("../input/input05.txt")
	ss, _ := AH.ParseLineGroups("../input/input05.txt", "|")

	mss := [][]Mapping{}
	for _, s := range ss[1:] {
		mss = append(mss, parseMapping(s))
	}

	seeds := parseSeeds(inputLines[0])

	part1 := math.MaxInt64
	for _, s := range seeds {
		for _, ms := range mss {
			s = applyMappingToInt(s, ms)
		}
		if (part1 > s) {
			part1 = s
		}
	}

	seedRanges := parseSeedsToRanges(inputLines[0])
	for _, ms := range mss {
		seedRanges = applyMappingToRanges(seedRanges, ms)
	}

	part2 := math.MaxInt64
	for _, r := range seedRanges {
		if part2 > r.s {
			part2 = r.s
		}
	}

	AH.PrintSoln(5, part1, part2)

	return
}
